### ISSUE TITLE
Output data as vector tiles

### ISSUE THREAD

**Issue by mstahv** [rphlo/karttapullautin#116]
Raster maps are so 90s ;-) https://github.com/mapbox/awesome-vector-tiles

Seriously, quite recently dropped usage of karttapullautin data in one of my hobby apps, because the [vector data via NLS](https://www.maanmittauslaitos.fi) is much smoother and clearer in browsers. Now if I only could get improved countours (and maybe improved vegetation as well) via karttapullautin combined with other vector data from fnls...

**Comment by stefankinell** [comment]
@mstahv - you do know that you already get Karttapullautin output in vector format? All from the vegetation files. And in this project the vegetation can now be delivered in bit-format images, which is really really easy to transform to vector format.

**Comment by mstahv** [comment]
Yes, even the older Perl version did vector format, but at least I haven't heard/read about vector **tiles**. And then of course a service like mapant.fi should host those 🧸

**Comment by mstahv** [comment]
Like I filled the issue ;-) The de-facto standard (MapBox Vector Tile Specification) needs to be the end result, geojson could be a middle format some other tooling can then parse. Might be though that geojson is not an optimal format there, haven't spent a minute to investigate 🤷‍♂️

**Comment by mstahv** [comment]
BTW. Do you know if the mapant tooling is OSS?

**Comment by NicoRio42** [comment]
@mstahv I am currently working on mapant.fr V2 as the LiDAR data are now broadly available in France. I am considering generating vector tiles from Karttapullautin output for contours and human features.
However I think it is out of the scope of Karttapullautin. In my opinion it should be part of the tile pyramid generation pipeline. I will publish the repo for the generation pipeline of mapant.fr V2 when I start working on it (in the coming weeks). So stay tuned if you are interested!

The python scripts in this [this repo](https://github.com/NicoRio42/mapant-scripts) are meant to generate a raster tiles pyramid from Karttapullautin output. However I don't guarantee the quality of the code as I was way less experimented back then 😉

**Comment by grst** [comment]
FWIW, on the note of building tile pyramids, I built [karttapullautin2tiles](https://github.com/grst/karttapullautin2tiles) that generates a web mercator tile directory directly from a karttapullautin `out` directory. 

But it's currently also about pixel, not vector, tiles.

**Comment by grst** [comment]
I'd like to try adding support for geoJSON. This should make it straightforward to generate vector tiles with tools such as [tippecanoe](https://github.com/felt/tippecanoe) it should be straightforward to generate vector tiles. If you don't mind, I'll send a draft PR to discuss details.

**Comment by grst** [comment]
On a second thought, [flatgeobuf](https://flatgeobuf.org/) might even be the better format. geoJSON according to [RFC7946](https://www.rfc-editor.org/info/rfc7946/) supports WGS84 projection only. But it would be nice if the output used the native projection of the LAZ tiles, also for consistency with the other outputs. Flatgeobuf is also the more efficient storage format and while not as ubiquitously supported as geojson, it is supported by a wide range of tools, including tippecanoe. 

what do you think?

**Comment by rphlo** [comment]
The ground work for geojson export has been made by @malpou in his fork. please check it out before starting anything new.

**Comment by grst** [comment]
ok, so I took a took at @malpou's code and it already does a lot of things, including vectorization of the vegetation layers. 

Before moving on, I'd like to take a step back and reflect on what outputs are really needed for a vector map. In particular, **I would appreciate if the maintainers could comment on what's in scope for karttapullautin and what should rather go to a downstream project** (e.g. I could add corresponding functionality to [karttapullautin2tiles](https://github.com/grst/karttapullautin2tiles)). 

## Overview of required data 

| Layer | currently available |sufficient |
|--|--|--|
| terrain (contours, dotknolls, formlines) | in `dxf` format |yes, geojson/flatgeobuf possibly better interoperabilty |
| vegetation | as bitmasks | needs vectorization (could be done downstream) |
| undergrowth | as bitmask | broken, see #275. Other than that same as vegetation | 
| OSM shapes | no | for mapant-scale maps, might be anyway easier to redraw them from the OSM source data than piping them through karttapullautin | 
| water | only rendered png | no, add bitmask as for vegetation or vector output | 
| blocks | only rendered png | no, add bitmask as for vegetation or vector output |  

## Way forward

If the answer is that vectorization should live downstream, then I think all that's needed is to ensure that all non-terrain layers are consistently exported as bitmasks. If vectorization should be done in karttapullautin, I could start working on breaking up @malpou's changes into smaller pieces such that they can be reviewed more easily.

**Comment by malpou** [comment]
> ok, so I took a took at @malpou's code and it already does a lot of things, including vectorization of the vegetation layers. 
> 
> Before moving on, I'd like to take a step back and reflect on what outputs are really needed for a vector map. In particular, **I would appreciate if the maintainers could comment on what's in scope for karttapullautin and what should rather go to a downstream project** (e.g. I could add corresponding functionality to [karttapullautin2tiles](https://github.com/grst/karttapullautin2tiles)). 
> 
> ## Overview of required data 
> 
> | Layer | currently available |sufficient |
> |--|--|--|
> | terrain (contours, dotknolls, formlines) | in `dxf` format |yes, geojson/flatgeobuf possibly better interoperabilty |
> | vegetation | as bitmasks | needs vectorization (could be done downstream) |
> | undergrowth | as bitmask | broken, see #275. Other than that same as vegetation | 
> | OSM shapes | no | for mapant-scale maps, might be anyway easier to redraw them from the OSM source data than piping them through karttapullautin | 
> | water | only rendered png | no, add bitmask as for vegetation or vector output | 
> | blocks | only rendered png | no, add bitmask as for vegetation or vector output |  
> 
> ## Way forward
> 
> If the answer is that vectorization should live downstream, than I think all that's needed is to ensure that all non-terrain layers are consistently exported as bitmasks. If vectorization should be done in karttapullautin, I could start working on breaking up @malpou's changes into smaller pieces such they can be reviewed more easily. 

I've basically tweaked it so I get vector outputs out which is then injected into a PostGIS DB. 

Then I'm using https://martin.maplibre.org/ as the map tile server.

Where I've made a ISOM 2017-2 styling package for the vector data doing some decent renderings.


Let me know if there's something I can contribute or help with, have ended up spending a lot of time getting the generation code to fit into more modern workflows and rendering.

Also done a whole thing of doing simulations on an model from the generated terrains.

**Comment by malpou** [comment]
Would love to collaborate / contribute, to making the world of auto generated maps more modern and create more complete pipelines that allows us to make the data more portable and usable in other scenarios and systems.

**Comment by malpou** [comment]
If anyone want a showcase (and is interested in it) of what's actually been done in the full pipeline I can cook up a write-up / blog post with that. Just let me know, it's currently just laying around in a private repo.

**Comment by rphlo** [comment]
Hi @malpou! Your fork changes have already been noticed and some improvements already merged in (slope lines things and compiler flags fine tuning).

I was delaying merging your vectorization commits, now that you are here we can talk about you contributing directly with some PRs.

@grst The bitmasks files were a step to allow vectorization downstream. Not sure if it has been used to that effect or…

@maplou code seemed logical and if he has proof of working cases it is probably good enough for the rest of us. I am doubtful continuing the bitmasks things would be the way to go.

Finally @malpou do you have some vector tile maps, generated from your branch, you could share with us to see what we could expect?

**Comment by malpou** [comment]
@rphlo I'll get back here once I've deployed the full vector tile on a server with PostGIS and Martin on a hostname, a bit busy with work these days. So haven't really touched the project I've been playing with for a month or something.

If anyone in the meantime wanna see some of the actual outcome of the things I've done. You can find an `.dxf` and a `.ocdCrt` [HERE ](https://drive.proton.me/urls/N0VBRJ6HK8#3N4yIO2Rtjde)for easy importing and mapping to ISOM symbols, I'm carrying the layers all the way through.

The data is not only the Lidar based generated maps, but is also paired with Linear & Polygon data from OpenStreetMap and Geodanmark for roads, houses, paths and what other layers i could find that provided value and have proper ISOM symbols.

Let me know if you have any questions?

I'm open for open sourcing the parts that are wrapped around `karttapullautin` to build this data pipeline, it's pretty extensible w. different other data sources that are available in other countries. All of that code is written in [go](https://go.dev/) on top of [PostGIS](https://postgis.net/). And it's at a point where i can draw a polygon on a map and i have a full vectorized map in about 15-60 minutes (injected to the database and ready for the `.dxf` export or viewable on a public map) on my laptop depending on size and data density.

Image from OCAD after a 1 click import and no treatment of the data coming out of my pipeline:
<img width="2443" height="1191" alt="Image" src="https://github.com/user-attachments/assets/da2621bd-85d3-4840-aaf9-32e736bd1aa6" />
[56.09661291576458, 9.699705250844705](https://maps.app.goo.gl/kVTbJUSZDCRbrXcU8)

**Comment by grst** [comment]
Hi @malpou, thanks for sharing, this sounds pretty neat! I'd have two questions regarding your workflow: 
 * do you directly load the geojson produced by your kp fork into postGIS and that's it, or are there any additional processing steps? 
 * do you send the OSM data through karttapullautin, or do you add it directly to postGIS? 

Regarding the implementation in your fork, it works quite nicely for me, but it currently doesn't export dotknolls, formlines, or cliffs as geojson, or am I missing something? I think it could also be nice to export the original greenshade in addition to the isom symbol, in case someone wants to use more fine-grained greens downstream.

**Comment by rphlo** [comment]
@malpou the first step would be for you to re-base your fork on latest master.
